helix-term/src/ui/document.rs

```diff
@@ -417,28 +396,65 @@ impl<'a> TextRenderer<'a> {
         } else {
             &self.tab
         };
-        let grapheme = match grapheme {
+        let mut whitespace_kind = WhitespaceKind::None;
+        let grapheme_value = match grapheme {
             Grapheme::Tab { width } => {
+                whitespace_kind = WhitespaceKind::Tab(width);
                 let grapheme_tab_width = char_to_byte_idx(tab, width);
                 &tab[..grapheme_tab_width]
             }
             // TODO special rendering for other whitespaces?
-            Grapheme::Other { ref g } if g == " " => space,
-            Grapheme::Other { ref g } if g == "\u{00A0}" => nbsp,
+            Grapheme::Other { ref g } if g == " " => {
+                whitespace_kind = WhitespaceKind::Space;
+                space
+            }
+            Grapheme::Other { ref g } if g == "\u{00A0}" => {
+                whitespace_kind = WhitespaceKind::NonBreakingSpace;
+                nbsp
+            }
             Grapheme::Other { ref g } => g,
-            Grapheme::Newline => &self.newline,
+            Grapheme::Newline => {
+                whitespace_kind = WhitespaceKind::Newline;
+                &self.newline
+            }
         };
 
-        let in_bounds = self.col_offset <= position.col
-            && position.col < self.viewport.width as usize + self.col_offset;
+        self.trailing_whitespace_tracker
+            .track(position.col, whitespace_kind);
+
+        let viewport_right_edge = self.viewport.width as usize + self.col_offset - 1;
+        let in_bounds = self.col_offset <= position.col && position.col <= viewport_right_edge;
 
         if in_bounds {
-            self.surface.set_string(
-                self.viewport.x + (position.col - self.col_offset) as u16,
-                self.viewport.y + position.row as u16,
-                grapheme,
-                style,
-            );
+            if self.trailing_whitespace_tracker.is_enabled()
+                && (grapheme == Grapheme::Newline || position.col == viewport_right_edge)
+            {
+                if let Some((from, trailing_whitespace)) = self.trailing_whitespace_tracker.get() {
+                    let offset = if from < self.col_offset {
+                        0
+                    } else {
+                        from - self.col_offset
+                    };
+                    let begin_at = if from < self.col_offset {
+                        self.col_offset - from
+                    } else {
+                        0
+                    };
+                    self.surface.set_string(
+                        self.viewport.x + offset as u16,
+                        self.viewport.y + position.row as u16,
+                        &trailing_whitespace[char_to_byte_idx(&trailing_whitespace, begin_at)..],
+                        style,
```

Comment:
This will messuo diagnostic and selection highkifjt s the style can frequently differ, you should only be updating the text while leaving cell style unchanged

Reply:
How can I render something without being forced to provide a style?

Reply:
I'm still not sure how to apply this feedback. I've changed the code to always apply `whitespace_style` though. I think this is basically the last pending thing to address.